LonaStudio/Controls/Picker/PickerView.swift

```diff
@@ -0,0 +1,152 @@
+//
+//  PickerView.swift
+//  LonaStudio
+//
+//  Created by Nghia Tran on 12/24/17.
+//  Copyright © 2017 Devin Abbott. All rights reserved.
+//
+
+import Cocoa
+
+class PickerView<Element: PickerItemType>: NSView {
+    
+    enum Option {
+        case placeholderText(String)
+        case selected(String)
+        case data([Element])
+        case viewForItem((NSTableView, Element, Bool) -> PickerRowViewType)
+        case didSelectItem((PickerView<Element>?, Element) -> Void)
+        case sizeForRow((Element) -> NSSize)
+    }
+    
+    struct Options {
+        var placeholderText: String = "Search ..."
+        var selected: String!
+        var data: [Element]!
+        var viewForItem: ((NSTableView, Element, Bool) -> PickerRowViewType)!
+        var didSelectItem: ((PickerView<Element>?, Element) -> Void)!
```

Comment:
If we put on required func on initial, it could be increase the complexity of initialization in Picker View.
I suggest to keep the Options 
```swift
enum Option {
    case placeholderText(String)
    case sizeForRow((Element) -> NSSize)
}
```
And introducing new `DataSource` (or we could find a better name)
```swift
enum DataSource {
    case selected(String)
    case data([Element])
    case viewForItem((NSTableView, Element, Bool) -> PickerRowViewType)
    case didSelectItem((PickerView<Element>?, Element) -> Void)
}
```
Then the initialization looks like
```swift
init(dataSource: [DataSource], options: [Option]) {}
```
How do you think 🤔

Reply:
That seems pretty reasonable. If we go with that, I'd probably make it:

```swift
struct Parameters {
    let selected: String
    let data: [Element]
    let viewForItem: (NSTableView, Element, Bool) -> PickerRowViewType
    let didSelectItem: (PickerView<Element>?, Element) -> Void
}
```

Although, if there are a lot of _required_ parameters I actually don't think it's bad to have a complex initializer. The caller will have to supply all of the parameters somehow, so I don't think there's a big difference between using a struct/enum and using individual parameters.

If we want to simplify the initializer, another option would be to look for parameters which might not need to be required. E.g. maybe `selected` and `didSelectItem` can be optional. The component still works correctly if `didSelectItem` defaults to `{_ in}`. And if it's possible to have nothing selected, then selected could be a nullable `String?`.

```swift
enum Option {
    case selected(String?)
    case placeholderText(String)
    case sizeForRow((Element) -> NSSize)
    case didSelectItem((PickerView<Element>?, Element) -> Void)
}

init(data: [Element], viewForItem: ..., options: [Option]) {

}
```

If our options are:
1. Create an initializer with 4 parameters + `Options` parameter
2. Create an initializer with `Parameters` struct + `Options` parameter
3. Make a few parameters optional, and create an initializer with 2 parameters + `Options` parameter

I think any of these options are fine. I slightly prefer (1) because it best represents how the component should be used. (2) adds an extra layer of indirection but ultimately the caller still has to define all the parameters. (3) could be fine, except that we shouldn't make parameters optional if we're currently _always_ passing them when we instantiate the component.

Reply:
Thank for clear explanation. I will follow the 1st options. It's good to me now 👍